MIG-1620: Virtual machine storage migration (#81461)
openshift/openshift-docs: modules/mtc-vm-creating-mig-plan.adoc

Reviewer: I don't really want users modifying the yaml themselves, there is a UI that does this.

Author: @awels - thanks very much for your input

Reviewer: Yes I would normally expect users to deploy VMs using the UI. The thing is it is a different UI (working on unifying this, but won't happen until at least 4.19) and I didn't want to modify the existing MTC documentation too much. If that is something that we think is important, we can spend some time writing down the steps on how create a migration plan with the UI.

Reviewer 2: I might not be fully understand what's going on here, so apologies if I'm missing some context :) 

In general, we would **especially** want to make UI steps very explicitly clear if they are to be completed in multiple UIs, and particularly if some steps are in a different UI than the user is used to using.

// Module included in the following assemblies:
// * migration_toolkit_for_containers/mtc-migrating-vms.adoc

:_mod-docs-content-type: PROCEDURE
[id="mtc-vm-creating-mig-plan_{context}"]
= Creating the migration plan

.Procedure

> . To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.

Author: @awels - please could you help clarify?

> . To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.

Reviewer: Its on line 24 here. One can migrate from a single namespace or multiple namespaces.

. Modify the contents of the migration plan by adding `mig-vm` to the namespaces.

> +
> .Example YAML

Author: ```suggestion
+
.Example migration plan YAML
```

+
[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  namespaces:
  - mig-vm # <1>
...
----
<1> add `mig-vm` to the namespaces.

> * In order to attempt a live storage migration, the `liveMigrate` field in the migration plan specification must be set to true, and KubeVirt must be configred and be enabled to perform live storage migration.

Author: ```suggestion
* In order to attempt a live storage migration, the `liveMigrate` field in the migration plan specification must be set to true, and `KubeVirt` must be configred and be enabled to perform live storage migration.
```

+
[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  liveMigrate: true # <2>
  namespaces:
...
----

> <2> Live migration only happens during the cutover of a migration plan. Staging the migration plan skips any running virtual machines and does not sync the data. Any stopped virtual machine disks will be synced.

Author: ```suggestion
<2> Live migration only happens during the cutover of a migration plan. 

Staging the migration plan skips any running virtual machines and does not sync the data. Any stopped virtual machine disks will be synced.
```

// Module included in the following assemblies:
//

:_mod-docs-content-type: PROCEDURE
[id="mtc-vm-creating-mig-plan_{context}"]
= Creating the migration plan

> To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.

Author: ```suggestion
.Procedure

* To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.
```

Modify the contents of the migration plan by adding `mig-vm` to the namespaces.

.Example YAML

[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  namespaces:
  - mig-vm # <1>
...
----
<1> add `mig-vm` to the namespaces.

In order to attempt a storage live migration the `liveMigrate` field in the spec must be set to true (and KubeVirt must be configred and be enabled to perform storage live migration.

[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  liveMigrate: true
  namespaces:
[... snipped, see config/samples/mig-plan.yaml for other required fields ...]
----

Live migration only happens during the cutover of a migration plan. Staging the migration plan skips any running virtual machines and does not sync the data. Any stopped virtual machine disks will be synced.

During the migration, a `MigMigration` resource is created indicating what type of migration is happening:

* stage
* rollback
* cutover

> 

Reviewer 3: Lines 51 forward: Maybe present simple instead of future?

Reviewer 4: Do we have more detailed explanation for what is stage/rollback/cutover?

Reviewer 5: > Do we have more detailed explanation for what is stage/rollback/cutover?

@duyanyan , this from _[Table 1. MTC terminology](https://docs.openshift.com/container-platform/4.16/migration_toolkit_for_containers/about-mtc.html#migration-terminology_about-mtc)_?
![image](https://github.com/user-attachments/assets/db13deb5-2693-4170-9e78-36c60568dd4b)

Author: @rayfordj - add an xref to this table? WDYT?

Reviewer 5: > @rayfordj - add an xref to this table? WDYT?

@anarnold97 , I'm inclined to say, "_yes_", but this is an area of docs where I'm ignorant.  Referencing the existing table makes sense and I think if using 'xref' is the way we do that here then, "yes, add an xref to this table" :sweat_smile:

> The status of the `MigMigration` will contain progress information about any storage live migrations.

Author: ```suggestion
The status of the `MigMigration` contains progress information about any storage live migrations.
```

// Module included in the following assemblies:

> //

Author: ```suggestion
// Module included in the following assemblies:
// * migration_toolkit_for_containers/mtc-migrating-vms.adoc
```

:_mod-docs-content-type: PROCEDURE
[id="mtc-vm-creating-mig-plan_{context}"]
= Creating the migration plan

.Procedure

* To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.

Modify the contents of the migration plan by adding `mig-vm` to the namespaces.

.Example YAML

[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  namespaces:
  - mig-vm # <1>
...
----
<1> add `mig-vm` to the namespaces.

> In order to attempt a storage live migration the `liveMigrate` field in the spec must be set to true (and KubeVirt must be configred and be enabled to perform storage live migration.

Author: ```suggestion
In order to attempt a storage live migration, the `liveMigrate` field in the spec must be set to true, and KubeVirt must be configred and be enabled to perform storage live migration.
```

// Module included in the following assemblies:
// * migration_toolkit_for_containers/mtc-migrating-vms.adoc

:_mod-docs-content-type: PROCEDURE
[id="mtc-vm-creating-mig-plan_{context}"]
= Creating the migration plan

> 

Author: @awels - please can you help with the prerequisites for the migration plan?

> 

Author: @awels - please could you help with what the user needs to know about creating the migration plan, and with the prerequisites?

> 

Reviewer: So there is a UI that comes with this that creates and manages the MigrationPlan. This is purely to help the user understand what the fields mean in the plan. I would not expect users to create/modify these themselves.

.Procedure

. To migrate the `mig-vm` namespace, ensure that the `namespaces` field of the migration plan includes `mig-vm`.

> +
> Modify the contents of the migration plan by adding `mig-vm` to the namespaces.

Author: ```suggestion
. Modify the contents of the migration plan by adding `mig-vm` to the namespaces.
```

+
.Example YAML
+
[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  namespaces:
  - mig-vm # <1>
...
----
+
<1> add `mig-vm` to the namespaces.

* In order to attempt a live storage migration, the `liveMigrate` field in the migration plan specification must be set to true, and KubeVirt must be configred and be enabled to perform live storage migration.
+
[source,yaml]
----
apiVersion: migration.openshift.io/v1alpha1
kind: MigPlan
metadata:
  name: live-migrate-plan
  namespace: openshift-migration
spec:
  liveMigrate: true
  namespaces:

> [... snipped, see config/samples/mig-plan.yaml for other required fields ...]

Author: ```suggestion
...
```

----

> * Live migration only happens during the cutover of a migration plan. Staging the migration plan skips any running virtual machines and does not sync the data. Any stopped virtual machine disks will be synced.

Reviewer 6: I don't think that lines 46-58 belong in this procedure module. Maybe they belong in a separate module that highlights what goes on "behind the scenes", like what I did here. https://docs.openshift.com/container-platform/4.16/networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.html#how-the-live-migration-process-works_migrate-from-openshift-sdn
